src/samVisualize/rendering/forceDirectedGraph.ts

```diff
@@ -0,0 +1,703 @@
+                            RenderConstants.secondaryTextSize * RenderConstants.mouseOverNodeSizeMultiplier
+                        )
+                }
+            )
+            .on(
+                'mouseout',
+                function (this: SVGGElement, event: d3.D3DragEvent<SVGGElement, NodeDatum, unknown>, d: NodeDatum) {
+                    d3.select(this)
+                        // Select the clipPath for the current node
+                        .select(`#clipCircle-${d.name}`)
+                        .select('circle')
+                        .transition()
+                        .duration(RenderConstants.mouseOverNodeGrowthTime)
+                        .attr('r', RenderConstants.radius)
+
+                    d3.select(this)
+                        .select('image')
+                        .transition()
+                        .duration(RenderConstants.mouseOverNodeGrowthTime)
+                        .attr('width', RenderConstants.radius * 2)
+                        .attr('height', RenderConstants.radius * 2)
+                        .attr('x', -RenderConstants.radius)
+                        .attr('y', -RenderConstants.radius)
+
+                    d3.select(this)
+                        .select('.primary-text')
+                        .transition()
+                        .duration(RenderConstants.mouseOverNodeGrowthTime)
+                        .attr('dy', RenderConstants.primaryLabelYOffset)
+                        .attr('font-size', RenderConstants.primaryTextSize)
+
+                    d3.select(this)
+                        .select('.secondary-text')
+                        .transition()
+                        .duration(RenderConstants.mouseOverNodeGrowthTime)
+                        .attr('dy', RenderConstants.secondaryLabelYOffset)
+                        .attr('font-size', RenderConstants.secondaryTextSize)
+                }
+            )
+        // vscodeApi does not exist outside of a Webview, so it's not used during testing
+        if (vscodeApi) {
+            nodes.on('click', function (event, d: NodeDatum) {
+                // Prevent click behavior on drag
+                if (event.defaultPrevented) {
+                    return
+                }
+                vscodeApi.postMessage({
+                    data: d.name,
+                })
+            })
+        }
+
+        // Define a uniquely identified clipPath to crop icons into circles,
+        // and allow each to be individually resized on mouseover.
+        nodes
+            .append<SVGClipPathElement>('clipPath')
+            .attr('id', (d: NodeDatum) => `clipCircle-${d.name}`)
+            .append<SVGCircleElement>('circle')
+            .attr('r', RenderConstants.radius)
+
+        // Draw an image corresponding to the Node type
+        nodes
+            .append<SVGImageElement>('image')
+            .attr('xlink:href', (d: NodeDatum) => {
+                const key = d.type?.toLowerCase()
```

Comment:
The icon selection logic is well-defined and shouldn't leak to the graph logic. Better to refactor it away.

Reply:
Will refactor to separate function.